File: packages/ai/src/usage.ts
import type { LlmProvider, ModelRole } from '@konusbitr/shared';

/**
 * Token and cost accounting, emitted once per model call.
 *
 * Phase 15 points this at Langfuse; Phase 13 bills on it. Both of those need
 * the same record, which is why it is collected from the start rather than
 * added when something consumes it: a usage series that begins the day
 * observability ships is a usage series with no history to compare against.
 *
 * What is deliberately *not* here: any of the text. A usage record is written
 * to logs an operator reads and shipped to a third-party observability
 * platform, and document text is untrusted input that must reach neither. So
 * the record carries counts and identifiers, and the closest it comes to
 * content is how many items were in the batch.
 */
export type UsageRecord = {
  role: ModelRole;
  provider: LlmProvider;
  model: string;
  /** Items in this call: passages embedded, messages sent, documents reranked. */
  items: number;
  promptTokens: number;
  completionTokens: number;
  /** Wall clock for the call, retries included. */
  durationMs: number;
  /** Attempts spent, the first included. */
  attempts: number;
  /**
   * Estimated cost in USD, or `null` when the model has no published price —
   * which is every local model, where the honest answer is "no marginal cost"
   * rather than zero dollars of a metered spend.
   */
  costUsd: number | null;
};

export type UsageSink = (record: UsageRecord) => void;

/**
 * Published prices, in USD per million tokens.
 *
 * A small hand-maintained table rather than a dependency: it covers the models
 * `DEFAULT_*_MODELS` names, it is only ever an estimate for a dashboard, and a
 * package that tracks provider pricing is a package that has to be updated on
 * somebody else's release schedule. A model that is not here reports `null`,
 * which renders as "unpriced" rather than as free.
 */
const PRICE_PER_MILLION_TOKENS: Readonly<Record<string, { input: number; output: number }>> =
  Object.freeze({
    'text-embedding-3-large': { input: 0.13, output: 0 },
    'text-embedding-3-small': { input: 0.02, output: 0 },
    'mistral-embed': { input: 0.1, output: 0 },
  });

export function estimateCostUsd(
  model: string,
  promptTokens: number,
  completionTokens: number,
): number | null {
  const price = PRICE_PER_MILLION_TOKENS[model];
  if (price === undefined) return null;
  return (promptTokens * price.input + completionTokens * price.output) / 1_000_000;
}

/**
 * The default sink: one structured line per call.
 *
 * `console` rather than a logger dependency, because this package is imported
 * by both the Next.js server and the scripts in `evals/`, and neither should
 * inherit a logging framework from a model router.
 */
export const logUsage: UsageSink = (record) => {
  // biome-ignore lint/suspicious/noConsole: emitting is this sink's only job
  console.info('[ai] model call', record);
};

/** Collect records instead of emitting them. Used by tests and by the evals. */
export function collectUsage(): { sink: UsageSink; records: UsageRecord[] } {
  const records: UsageRecord[] = [];
  return { sink: (record) => void records.push(record), records };
}

File: services/worker/src/konusbitr_worker/ai/usage.py
"""Token and cost accounting, emitted once per model call.

Phase 15 points this at Langfuse and Phase 13 bills on it. Both need the same
record, which is why it is collected from the start rather than added when
something consumes it: a usage series that begins the day observability ships
is a usage series with nothing to compare against.

What is deliberately absent is any of the text. A usage record goes to logs an
operator reads and, later, to a third-party platform — and document text is
untrusted input that must reach neither. So the record carries counts and
identifiers, and the closest it comes to content is how many items were in the
batch.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass

from konusbitr_worker.log import get_logger
from konusbitr_worker.settings import LlmProvider, ModelRole

__all__ = ["UsageRecord", "log_usage"]

logger = get_logger("konusbitr.worker.ai.usage")


@dataclass(frozen=True, slots=True)
class UsageRecord:
    """One model call, as accounting sees it."""

    role: ModelRole
    provider: LlmProvider
    model: str
    #: Items in this call: passages embedded, messages sent, documents reranked.
    items: int
    prompt_tokens: int
    completion_tokens: int
    #: Wall clock for the call, retries included.
    duration_ms: int
    #: Attempts spent, the first included.
    attempts: int
    #: Estimated USD, or ``None`` when the model has no published price — which
    #: is every local model, where the honest answer is "no marginal cost"
    #: rather than zero dollars of a metered spend.
    cost_usd: float | None

    def to_json(self) -> dict[str, object]:
        return asdict(self)


def log_usage(record: UsageRecord) -> None:
    logger.info("model call", extra={"usage": record.to_json()})
